src/elements/element.rectangle.js

```diff
@@ -78,70 +83,92 @@ module.exports = Element.extend({
 			bottom = vm.y + vm.height / 2;
 			signX = right > left ? 1 : -1;
 			signY = 1;
-			borderSkipped = vm.borderSkipped || 'left';
-		}
-
-		// Canvas doesn't allow us to stroke inside the width so we can
-		// adjust the sizes to fit if we're setting a stroke on the line
-		if (borderWidth) {
-			// borderWidth shold be less than bar width and bar height.
-			var barSize = Math.min(Math.abs(left - right), Math.abs(top - bottom));
-			borderWidth = borderWidth > barSize ? barSize : borderWidth;
-			var halfStroke = borderWidth / 2;
-			// Adjust borderWidth when bar top position is near vm.base(zero).
-			var borderLeft = left + (borderSkipped !== 'left' ? halfStroke * signX : 0);
-			var borderRight = right + (borderSkipped !== 'right' ? -halfStroke * signX : 0);
-			var borderTop = top + (borderSkipped !== 'top' ? halfStroke * signY : 0);
-			var borderBottom = bottom + (borderSkipped !== 'bottom' ? -halfStroke * signY : 0);
-			// not become a vertical line?
-			if (borderLeft !== borderRight) {
-				top = borderTop;
-				bottom = borderBottom;
-			}
-			// not become a horizontal line?
-			if (borderTop !== borderBottom) {
-				left = borderLeft;
-				right = borderRight;
-			}
+			borderSkipped = valueOrDefault(vm.borderSkipped, 'left') || '';
 		}
 
-		ctx.beginPath();
 		ctx.fillStyle = vm.backgroundColor;
 		ctx.strokeStyle = vm.borderColor;
-		ctx.lineWidth = borderWidth;
-
-		// Corner points, from bottom-left to bottom-right clockwise
-		// | 1 2 |
-		// | 0 3 |
-		var corners = [
-			[left, bottom],
-			[left, top],
-			[right, top],
-			[right, bottom]
-		];
-
-		// Find first (starting) corner with fallback to 'bottom'
-		var borders = ['bottom', 'left', 'top', 'right'];
-		var startCorner = borders.indexOf(borderSkipped, 0);
-		if (startCorner === -1) {
-			startCorner = 0;
+
+		if (!borderWidth) {
+			ctx.fillRect(left, bottom, right - left, top - bottom);
+			return;
 		}
 
-		function cornerAt(index) {
-			return corners[(startCorner + index) % 4];
+		maxWidth = Math.abs(left - right) / 2;
+		maxHeight = Math.abs(top - bottom) / 2;
+		if (helpers.isObject(borderWidth)) {
+			borderWidth = {
+				bottom: Math.min(borderWidth.bottom || 0, maxHeight),
+				left: Math.min(borderWidth.left || 0, maxWidth),
+				top: Math.min(borderWidth.top || 0, maxHeight),
+				right: Math.min(borderWidth.right || 0, maxWidth)
+			};
+		} else {
+			borderWidth = {
+				bottom: borderSkipped === 'bottom' ? 0 : Math.min(borderWidth, maxHeight),
```

Comment:
is it worth caching `Math.min(borderWidth, maxHeight)` and `Math.min(borderWidth, maxWidth)` into local vars?

Reply:
Done